[Design] 로그인 / 회원가입 UI

Gobans opened the pull request:
## 작업내용
<!-- 작업 내용과 이미지를 첨부해주세요. -->
- [x] 바뀐 디자인에 따라 UI 다시 작업

<!-- (+스크린샷)이 있다면 적어주세요. 없으면 지워주세요-->
### iPhone SE

<img width="250" height="450" src="https://user-images.githubusercontent.com/56781342/197327938-576bbabc-61dd-422f-b0e9-d3cc6ea9e0d0.png"><img width="250" height="450" src="https://user-images.githubusercontent.com/56781342/197327953-14571dd5-796e-4c15-8f15-46a5928c4059.png"><img width="250"    height="450" src="https://user-images.githubusercontent.com/56781342/197327997-889c1e2d-76ca-4b6f-932a-0dd67ee693df.png">

### iPhone 13
<img width="250" height="450" src="https://user-images.githubusercontent.com/56781342/197328143-25c8fb6a-c8f5-4dbc-a6d6-2c756edd356a.png"><img width="250" height="450" src="https://user-images.githubusercontent.com/56781342/197328151-186835c3-4c4b-4354-a769-fc42c0c6d6be.png"><img width="250"    height="450" src="https://user-images.githubusercontent.com/56781342/197328155-5739042d-e6bd-4f58-811f-6888fda56675.png">

### iPhone 13 Pro Max

<img width="250" height="450" src="https://user-images.githubusercontent.com/56781342/197328261-64f7deca-15ef-433d-a646-2d610483a88e.png"><img width="250" height="450" src="https://user-images.githubusercontent.com/56781342/197328270-202ca8bb-3e5f-480d-9832-c99054f9cba4.png"><img width="250"    height="450" src="https://user-images.githubusercontent.com/56781342/197328284-0d37bc82-f622-41e9-81df-e3c0da76b199.png">

### textField 마지막 글자 입력 시 받침도 입력 되도록 로직 추가
- 닉네임 입력시 마지막 글자가 하나 들어가면, 더이상 입력을 할 수 없는 부분을 로직을 추가하여 수정했습니다.
15acb80d94a4d5fcd33ba390dd095cdc16809350

### 공용 컴포넌트 ArrowButton
<img width="70" alt="image" src="https://user-images.githubusercontent.com/56781342/197329567-38dd4a15-a7bf-4b57-b6b6-b0870bddb50d.png">

- startIndicator와 stopIndicator로 내부에 있는 UIActivityIndicatorView를 동작시킬 수 있습니다.
- 코드안에 설명 있어요

### FullWidthBlackButton
- 설명 추가했습니다

## 리뷰포인트
<!-- 리뷰가 필요한 포인트와 해당 되는 커밋을 링크로 걸어주세요. -->
- 전체적인 디자인을 특히 @Chaeho-Min 가 봐주시면 좋을 것 같습니다.

## Checklist
- [x] 브랜치를 가져와 작업한 경우 이전 브랜치에 PR을 보냈는지 확인
- [x] 빌드를 위해 SceneDelegate 수정한 것 PR로 올리지 않았는지 확인
- [x] 필요없는 주석, 프린트문 제거했는지 확인
- [x] 컨벤션 지켰는지 확인
- [x] final, private 제대로 넣었는지 확인
- [x] 다양한 디바이스에 레이아웃이 대응되는지 확인
  - [x] iPhone SE
  - [x] iPhone 13
  - [x] iPhone 13 Pro Max

Gobans committed:
Design:art:: Shadow 버튼 UI 변경

ShadowButton -> ArrowButton 디자인 변경

Gobans committed:
Design:art:: signup login 이미지 변경

Gobans committed:
Design:art:: ThirdPartyLoginView 디자인 변경

Gobans committed:
Design:art:: FullWidthBlackButton  컴포넌트 생성

Gobans committed:
Design:art:: Full width Button 사용

Gobans committed:
Design:art:: ThirdPartyLoginView Constraints 변경

Gobans committed:
Design:art:: UI Constarints 변경

Gobans committed:
Setting:gear:: Merge with dev

Gobans committed:
Chore:broom:: Merge에 따른 Failed 코드 변경

Gobans committed:
Design:art:: Login background image 변경

Gobans committed:
Design:art:: MainTitleView 공용 컴포넌트 적용

Gobans committed:
Design:art:: NickNameInputView 공용 컴포넌트 적용

Gobans committed:
Design:art:: ThirdPartyLoginView constraints 변경

Gobans committed:
Design:art:: 약관 동의 Color 변경

Gobans committed:
Design:art:: UI 디자인 전체 수정

Gobans committed:
Feat:sparkles:: 마지막 입력글자 처리 로직 추가

#Fix:ladybug:: 버그 수정
#Docs:books:: 문서 수정
#Design:art:: UI 디자인 변경
#Resource:bento:: 리소스 추가
#Refactor:recycle:: 코드 리팩토링
#Chore:broom:: 자질구레한 수정사항
#Setting:gear:: 패키지 추가 등 세팅 관련
#HOTFIX:ambulance:: 치명적인 버그 수정

Gobans committed:
Chore:broom:: 글자 길이 변수 viewModel로 이전

Gobans committed:
Chore:broom:: 최대 닉네임 글자수 변수 이름 변경

#Fix:ladybug:: 버그 수정
#Docs:books:: 문서 수정
#Design:art:: UI 디자인 변경
#Resource:bento:: 리소스 추가
#Refactor:recycle:: 코드 리팩토링

#Setting:gear:: 패키지 추가 등 세팅 관련
#HOTFIX:ambulance:: 치명적인 버그 수정

Gobans committed:
Docs:books:: FullWidthBlackButton 설명 추가

Gobans committed:
#Docs:books:: ArrowButton 설명 추가

Chaeho-Min commented on Projects/App/Sources/UI/User/Login/View/ThirdPartyLoginViewController.swift:
```diff
@@ -52,51 +56,58 @@ extension ThirdPartyLoginViewController {
         kakaoLoginButton.addTarget(self, action: #selector(kakaoLoginButtonClicked), for: .touchUpInside)
         kakaoLoginButton.setImage(UIImage(.btn_kakaologin), for: .normal)
-        kakaoLoginButton.imageView?.contentMode = .scaleAspectFit
+        kakaoLoginButton.imageView?.contentMode = .scaleAspectFill
```
p1;
<img width="336" alt="image" src="https://user-images.githubusercontent.com/75792767/197409783-80902de7-d338-40a0-a2f5-53f43387e91e.png">
<img width="336" alt="image" src="https://user-images.githubusercontent.com/75792767/197409810-857d7f59-342a-43b4-b97f-e8ef21a547fc.png">

여기 scaleAspectFit에서 scaleAspectFill로 바뀌었는데 혹시 이유가 있나요?
이러면 이미지 버튼이 잘릴 수도 있어서 다시 Fit으로 바꿔주세요!

아 그리고 로그인 버튼 높이 55입니다! 이것도 수정해주세요

Gobans replied:
이 부분이 높이가 정해져 있기 때문에 scaleAspectFit 하면 비율에 맞게 길이가 충분히 길어지지 못하는 현상이 있어서 바꿨던 것 같습니다.
<img width="468" alt="image" src="https://user-images.githubusercontent.com/56781342/197439109-95c7cd51-8ec0-4532-91ef-ae21b837691f.png">
이렇게도 괜찮다면 그대로 하겠습니다! 높이는 수정하였습니다.
6c981779fe18fa285e0487b46150e66122dd3c01

Chaeho-Min commented on Projects/App/Sources/UI/User/Login/View/ThirdPartyLoginViewController.swift:
```diff
@@ -52,51 +56,58 @@ extension ThirdPartyLoginViewController {
         appleLoginButton.addTarget(self, action: #selector(appleLoginButtonClicked), for: .touchUpInside)
         appleLoginButton.setImage(UIImage(.btn_applelogin), for: .normal)
-        appleLoginButton.imageView?.contentMode = .scaleAspectFit
+        appleLoginButton.imageView?.contentMode = .scaleAspectFill
```
여기도 Fit으로!

Gobans replied:
일단 수정완료! 6c981779fe18fa285e0487b46150e66122dd3c01

Chaeho-Min commented on Projects/App/Sources/UI/User/Signup/View/NickNameInputViewController.swift:
```diff
@@ -16,11 +17,9 @@ final class NickNameInputViewController: UIViewController {
-    private let titleLabel = UILabel()
-    private let subtitleLabel = UILabel()
+    private let mainTitleView = MainTitleView(title: "닉네임을 알려주세요", subtitle: "언제든지 변경할 수 있어요.")
     private let nickNameTextField = UITextFieldPadding(top: 14, left: 20, bottom: 14, right: 20)
```
p2;

이 부분에 대한건 아닌데 리뷰 달 데가 없어서 여기에다 답니다ㅎㅎ

<img width="250" src="https://user-images.githubusercontent.com/75792767/197410012-3c5b3a9f-147f-4c3f-acd5-69671fb0f36f.gif">

뷰가 넘어오자마자 키보드가 올라오게 viewDidLoad에 nickNameTextField.becomeFirstResponder() 추가해주세요!

그리고 nickNameTextField가 mainTItleView와 키보드 사이 공간의 중앙에 위치했으면 좋겠어요 해주세요!😜
UIView 하나 만들어서 topAnchor를 mainTitleView.bottomAnchor에 맞추고, bottomAnchor를 키보드 상단에 맞춘다음에 nickNameTextField를 UIView의 중앙에 위치시키면 되지 않을까... 근데 이렇게 하면 키보드 올라갔다 내려갈때마다 버벅이니까 키보드 비활성화를 또 막아야 할듯요?

Lia316 replied:
그건 또 제가 완벽하게 해냈죠 😋
아보가 말씀하신 대로 하고, 애니메이션 추가하면 자연스럽긴 합니다!

tea-hkim replied:
확인했습니다!
@Chaeho-Min 대학교 도메인 입력도 이런 방식으로 처리하면 되겠죠?

Gobans replied:
<img width="250" height="500" alt="image" src="https://user-images.githubusercontent.com/56781342/197452886-de39b946-6779-4119-9057-ae679c90f78f.gif">
반영완료!
536c79574a3ec46a337b34e7a6353fae5373fe22

Chaeho-Min commented on Projects/App/Sources/UI/User/Signup/ViewModel/NickNameInputViewModel.swift:
```diff
@@ -52,9 +54,8 @@ final class NickNameInputViewModel {
     }
     
     func isValid(for input: String) -> Bool {
```
p1;
이것도 코드리뷰를 해당 부분에 달 수 없어서 여기에 답니다.
지금 닉네임에 공백을 입력할 수 있어서 이것도 막아주세요!

Gobans replied:
반영완료!
344c35b7e8227a62ccdad3d171bd8e81ea242416

Chaeho-Min commented on Projects/App/Sources/UI/User/Login/View/ThirdPartyLoginViewController.swift:
```diff
@@ -7,6 +7,7 @@
 //
 
 import AuthenticationServices
+import SwiftUI
```
p3;
디자인 관련해서만 리뷰 달려고 했는데 이거 하나만 더 달고 끝내겠습니다!
프리뷰에서만 SwiftUI를 쓰는데, 여기서 임포트하는건 뭔가 찜찜해서 아래처럼 디버깅시에만 임포트하게 하면 어떨까요?

```swift
#if DEBUG
import SwiftUI
struct ThirdPartyLoginViewTemplatePreview: PreviewProvider {
     static var previews: some View {
         ThirdPartyLoginViewController().toPreview()
     }
 }
#endif
```

Gobans replied:
이거 깜빡하고 있었네요. 
감사합니다 반영하였습니다!
fce0567bad183ea873db9e74aac71ce48cf18c28 3cc35d377c56038bd933524474d6d7095beaa301

Chaeho-Min replied:
근데 이거 보니까 템플릿이더라고요... 협의해서 템플릿을 바꾸던가 해야될듯요

Chaeho-Min reviewed (changes requested):
UI랑 컴포넌트 만드느라 정말 고생 많으셨습니다!
디자인 시안이랑 다른 부분이 좀 있어서 수정 요청 드렸으니 확인해주세요!

아 그리고 회원가입 플로우에서는 뒤로가기 다 막아야 할 것 같아요 막아주세요ㅎㅎ🙇‍♂️

tea-hkim commented on Projects/App/Sources/UI/User/Signup/View/SignupCompleteViewController.swift:
```diff
@@ -43,62 +39,48 @@ extension SignupCompleteViewController {
-        
-        startButtonView.button.setAttributedTitle(NSAttributedString(string: "시작하기",
-                                                                     attributes: [NSAttributedString.Key.font: UIFont.systemFont(ofSize: 17, weight: .bold)]), for: .normal)
+        startButton.setAttributedTitle(NSMutableAttributedString(string: "우리 대학 맛집여정에 함께하기",
```
Q. setTitle이 아니라 setAttributedTitle으로 하신 이유가 궁금합니다!

Gobans replied:
컴포넌트 바뀌면서 수정을 깜빡했네요 짚어주셔서 감사합니다!
21ab369fd994a8b843af7c48bf43003bcf233c4a

tea-hkim reviewed (approved):
깔끔하네요 역시!
아보가 말한 부분만 수정되면 될 것 같습니다!수고하셨습니다!

Gobans committed:
Design:art:: 로그인 버튼 수정

#Fix:ladybug:: 버그 수정
#Docs:books:: 문서 수정

#Resource:bento:: 리소스 추가
#Refactor:recycle:: 코드 리팩토링
#Chore:broom:: 자질구레한 수정사항
#Setting:gear:: 패키지 추가 등 세팅 관련
#HOTFIX:ambulance:: 치명적인 버그 수정

Gobans committed:
Design:art:: 키보드 show hide 디자인과 코드 변경

Gobans committed:
Chore:broom:: 닉네임 공백입력 불가처리

Gobans committed:
Chore:broom:: Preview If DEBUG 처리

#Fix:ladybug:: 버그 수정
#Docs:books:: 문서 수정
#Design:art:: UI 디자인 변경
#Resource:bento:: 리소스 추가
#Refactor:recycle:: 코드 리팩토링

#Setting:gear:: 패키지 추가 등 세팅 관련
#HOTFIX:ambulance:: 치명적인 버그 수정

Gobans committed:
Chore:broom:: import swiftUI 위치변경

Gobans committed:
Chore:broom:: 필요없는 setAttributeString 삭제

Gobans commented:
@Chaeho-Min @tea-hkim 
세심한 코드리뷰 감사합니다! 수정사항 반영해서 코멘트 달았는데 확인 부탁드립니다~!

Chaeho-Min commented on Projects/Network/Sources/EndPoint/Environment.swift:
```diff
@@ -0,0 +1,14 @@
+
+import Foundation
+
+struct Environment {
+//    static let ZESTY_HOST = "43.201.29.11"
+    static let PORT = 8080
+}
```
어 이거... 여기 PR에 올라와도 되는 건가요?

Gobans replied:
아 이거 올라오면 안되는데..ㅜ 확인 감사합니다 수정했습니다.
뒤로가기가 근데 닉네임 중복체크 후 바로 다음뷰로 넘어가서 되돌아가지 못해도 괜찮을까요?
뭔가 체크를 하고 사용자에게 표시한 후 그 다음에 다시 버튼을 눌러 닉네임 확정을 짓는 플로우가 필요해보입니다 (나중에)

Gobans replied:
72f72ae4cc82b597ea072670b4d06c0cc84baa00

Chaeho-Min replied:
버튼이 '중복 체크 버튼'이라고 표시되어있지 않고 그냥 화살표로 표시되어 있으니까 사용자에겐 닉네임을 확정짓고 다음으로 넘어가는 버튼으로 인식될거 같아서 괜찮을 것 같습니다.
그리고 중복 체크 버튼 누르고 -> 다음 버튼 누르게 되면 태스크가 하나 더 추가되는게 싫기도 했고요.
그런데 로직상 SignupCompleteViewController의 버튼을 눌렀을 때 회원가입이 완료된다면, Signup 뷰에서는 뒤로가기가 되어도 괜찮을 것 같긴 하네요.

Chaeho-Min reviewed (commented):
수정사항 반영 감사드립니다!
그런데 이번 수정사항 반영하면서 Environment.swift 파일이 같이 PR에 올라왔는데 이거 여기 PR에 올라오면 안되는거 아닌가요???
아 그리고 뒤로가기 아직 안 막힌거 같은데 이것도 막아주시면 감사...😅

Gobans committed:
Chore:broom:: 잘못들어간 enviroment 제거

Gobans committed:
Feat:sparkles:: navigation, swipe 로 이전 뷰 이동 막기

Chaeho-Min reviewed (approved):
🚀🚀🚀
